File: Data Structures & Algorithms/decode-ways/submission-3.py
class Solution:
    def numDecodings(self, s: str) -> int:
        n = len(s)
        last, secondLast = 1, 1
        for i in range(n - 1, -1, -1):
            if s[i] =='0':
                secondLast = last
                last = 0
            elif (
                i < n - 1
                and (
                    s[i] == '1' or
                    (s[i] == '2' and '0' <= s[i+1] <= '6')
                    )
                ):
                temp = last + secondLast
                secondLast = last
                last = temp
            else:
                secondLast = last
        
        return last

File: Data Structures & Algorithms/house-robber/submission-3.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        int last = 0, prev = 0;

        for(int i: nums) {
            int temp = max(prev + i, last);
            prev = last;
            last = temp;
        }

        return last;
    }
};
